hudi-common/src/main/java/org/apache/hudi/common/util/queue/DisruptorExecutor.java

```diff
@@ -0,0 +1,168 @@
+                           final String waitStrategy, Runnable preExecuteRunnable) {
+    this.producers = producers;
+    this.consumer = consumer;
+    this.preExecuteRunnable = preExecuteRunnable;
+    // Ensure fixed thread for each producer thread
+    this.producerExecutorService = Executors.newFixedThreadPool(producers.size(), new CustomizedThreadFactory("producer"));
+    // Ensure single thread for consumer
+    this.consumerExecutorService = Executors.newSingleThreadExecutor(new CustomizedThreadFactory("consumer"));
+    this.queue = new DisruptorMessageQueue<>(bufferSize, transformFunction, waitStrategy, producers.size(), preExecuteRunnable);
+  }
+
+  /**
+   * Start all Producers.
+   */
+  public ExecutorCompletionService<Boolean> startProducers() {
+    final ExecutorCompletionService<Boolean> completionService =
+        new ExecutorCompletionService<Boolean>(producerExecutorService);
+    producers.stream().map(producer -> {
+      return completionService.submit(() -> {
+        try {
+          preExecuteRunnable.run();
+
+          DisruptorPublisher publisher = new DisruptorPublisher<>(producer, queue);
+          publisher.startProduce();
+
+        } catch (Throwable e) {
+          LOG.error("error producing records", e);
+          throw e;
+        }
+        return true;
+      });
+    }).collect(Collectors.toList());
+    return completionService;
+  }
+
+  @Override
+  public E execute() {
+    try {
+      assert consumer.isPresent();
+      setupConsumer();
+      ExecutorCompletionService<Boolean> pool = startProducers();
+
+      waitForProducersFinished(pool);
+      queue.getInnerQueue().shutdown();
+      consumer.get().finish();
+
+      return consumer.get().getResult();
+    } catch (InterruptedException ie) {
+      shutdownNow();
+      Thread.currentThread().interrupt();
+      throw new HoodieException(ie);
+    } catch (Exception e) {
+      throw new HoodieException(e);
+    }
+  }
+
+  private void waitForProducersFinished(ExecutorCompletionService<Boolean> pool) throws InterruptedException, ExecutionException {
+    for (int i = 0; i < producers.size(); i++) {
+      pool.take().get();
+    }
+  }
+
+  /**
+   * Start only consumer.
+   */
+  private void setupConsumer() {
+    DisruptorMessageHandler<O, E> handler = new DisruptorMessageHandler<>(consumer.get());
+
+    Disruptor<HoodieDisruptorEvent<O>> innerQueue = queue.getInnerQueue();
+    innerQueue.handleEventsWith(handler);
+    innerQueue.start();
+  }
+
+  @Override
+  public boolean isRemaining() {
+    return !queue.isEmpty();
+  }
+
+  @Override
+  public void shutdownNow() {
+    producerExecutorService.shutdownNow();
+    consumerExecutorService.shutdownNow();
+    queue.close();
+  }
+
+  @Override
+  public DisruptorMessageQueue<I, O> getQueue() {
+    return queue;
+  }
+
+  @Override
+  public boolean awaitTermination() {
```

Comment:
+1

Reply:
also, why return true directly. should we check for producerExecutorService termination and consumerExecutorService termination. similar logic as followed in BoundedInMemoryExecutor

Reply:
> Please check my comment above: both BoundedInMemory and Disruptor Executors could share these sequences (for managing lifecycles of executor services)

changed!

> also, why return true directly. should we check for producerExecutorService termination and consumerExecutorService termination. similar logic as followed in BoundedInMemoryExecutor

Since we now have `HoodieExecutorBase` it holds common elements producerExecutorService, consumerExecutorService, producers and consumer and share the same logic for executor awaitTermination.